starky2/src/cross_table_lookup.rs

```diff
@@ -0,0 +1,299 @@
+use anyhow::{ensure, Result};
+use plonky2::field::extension_field::{Extendable, FieldExtension};
+use plonky2::field::field_types::Field;
+use plonky2::field::packed_field::PackedField;
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::challenger::Challenger;
+use plonky2::plonk::config::GenericConfig;
+use plonky2::plonk::plonk_common::reduce_with_powers;
+use plonky2::util::reducing::ReducingFactor;
+
+use crate::all_stark::Table;
+use crate::config::StarkConfig;
+use crate::constraint_consumer::ConstraintConsumer;
+use crate::permutation::{
+    get_grand_product_challenge_set, GrandProductChallenge, GrandProductChallengeSet,
+};
+use crate::proof::StarkProofWithPublicInputs;
+use crate::stark::Stark;
+use crate::vars::StarkEvaluationVars;
+
+#[derive(Clone)]
+pub struct CrossTableLookup<F: Field> {
+    pub looking_table: Table,
+    pub looking_columns: Vec<usize>,
+    pub looked_table: Table,
+    pub looked_columns: Vec<usize>,
+    pub default: Vec<F>,
+}
+
+impl<F: Field> CrossTableLookup<F> {
+    pub fn new(
+        looking_table: Table,
+        looking_columns: Vec<usize>,
+        looked_table: Table,
+        looked_columns: Vec<usize>,
+        default: Vec<F>,
+    ) -> Self {
+        assert_eq!(looking_columns.len(), looked_columns.len());
+        Self {
+            looking_table,
+            looking_columns,
+            looked_table,
+            looked_columns,
+            default,
+        }
+    }
+}
+
+/// Cross-table lookup data for one table.
+#[derive(Clone)]
+pub struct CtlData<F: Field> {
+    /// Challenges used in the argument.
+    pub(crate) challenges: GrandProductChallengeSet<F>,
+    /// Vector of `(Z, columns)` where `Z` is a Z-polynomial for a lookup on columns `columns`.
+    pub zs_columns: Vec<(PolynomialValues<F>, Vec<usize>)>,
+}
+
+impl<F: Field> CtlData<F> {
+    pub(crate) fn new(challenges: GrandProductChallengeSet<F>) -> Self {
+        Self {
+            challenges,
+            zs_columns: vec![],
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.zs_columns.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.zs_columns.is_empty()
+    }
+
+    pub fn z_polys(&self) -> Vec<PolynomialValues<F>> {
+        self.zs_columns.iter().map(|(p, _)| p.clone()).collect()
+    }
+}
+
+pub fn cross_table_lookup_data<F: RichField, C: GenericConfig<D, F = F>, const D: usize>(
+    config: &StarkConfig,
+    trace_poly_values: &[Vec<PolynomialValues<F>>],
+    cross_table_lookups: &[CrossTableLookup<F>],
+    challenger: &mut Challenger<F, C::Hasher>,
+) -> Vec<CtlData<F>> {
+    let challenges = get_grand_product_challenge_set(challenger, config.num_challenges);
+    cross_table_lookups.iter().fold(
```

Comment:
nit: likewise here, to me it seems a little harder to read a fold (vs a for loop) with mutations and a large body